vote/src/vote_account.rs

```diff
@@ -65,6 +74,74 @@ pub struct VoteAccounts {
     >,
 }
 
+impl VoteAccounts {
+    pub fn clone_and_filter_for_alpenglow(
+        &self,
+        max_vote_accounts: usize,
+        minimum_identity_account_balance: u64,
+        // The identity_account_balance is indexed with vote account pubkey.
+        // The value is the lamport balance of the identity account.
+        identity_account_balances: &HashMap<Pubkey, u64>,
+    ) -> VoteAccounts {
+        if max_vote_accounts == 0 {
+            panic!("max_vote_accounts must be > 0");
+        }
+        let mut entries_to_sort: Vec<(&Pubkey, &VoteAccount, u64)> = self
+            .vote_accounts
+            .iter()
+            .filter_map(|(pubkey, (stake, vote_account))| {
+                if vote_account
+                    .vote_state_view()
+                    .bls_pubkey_compressed()
+                    .is_some()
+                    && *stake != 0u64
+                {
+                    let identity_account_balance = identity_account_balances.get(pubkey)?;
+                    if *identity_account_balance < minimum_identity_account_balance {
+                        return None;
+                    }
+                    Some((pubkey, vote_account, *stake))
+                } else {
+                    None
+                }
+            })
+            .collect();
+        // Sort by stake descending
+        if entries_to_sort.len() > max_vote_accounts {
+            entries_to_sort.sort_by(|a, b| b.2.cmp(&a.2));
+            // Find the stake of the first one being in the truncated list (so it's not max_vote_accounts - 1)
+            let floor_stake = entries_to_sort.get(max_vote_accounts).unwrap().2;
+            entries_to_sort.truncate(max_vote_accounts);
+            // Per SIMD 357, we remove all vote accounts with stake equal to the first truncated one.
+            entries_to_sort.retain(|(_pubkey, _vote_account, stake)| *stake > floor_stake);
+        }
+        let valid_entries: HashMap<Pubkey, (u64, VoteAccount)> = entries_to_sort
+            .into_iter()
+            .map(|(pubkey, vote_account, stake)| (*pubkey, (stake, vote_account.clone())))
+            .collect();
+        if valid_entries.is_empty() {
+            panic!("no valid alpenglow vote accounts found");
+        }
+        VoteAccounts {
+            vote_accounts: Arc::new(valid_entries),
+            staked_nodes: OnceLock::new(),
+        }
+    }
+
+    pub fn identity_accounts_for_staked_nodes(&self) -> Vec<&Pubkey> {
+        self.vote_accounts
+            .iter()
+            .filter_map(|(_, (stake, vote_account))| {
+                if *stake != 0u64 {
```

Comment:
do we care about `bls_pubkey_compressed` here?

Reply:
It doesn't make a difference, this one is only for retrieving identity pubkeys from vote accounts so we can look up balances. I prefer to remove the `bls_pubkey_compressed` check here to keep the code simple, we will filter them out in `clone_and_filter` function above.

Reply:
yeah, I guess the implication is we potentially have a slightly larger hashmap with entries that we never ask for. In exchange we can eliminate an additional check. I'm fine either way

Reply:
With a simple lookup on 2000+ vote accounts I think it should be okay. Theoretically someone could spam us by adding vote accounts, but they could as easily add a BLS pubkey and stake 1 lamport, so I don't think it can help us much if we have intentional attackers.